server/etcdserver/api/v3health/healthz.go

```diff
@@ -0,0 +1,378 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3health
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/server/v3/auth"
+)
+
+const (
+	PathLivez   = "/livez"
+	PathReadyz  = "/readyz"
+	PathHealthz = "/healthz"
+)
+
+type EtcdServerHealth interface {
+	Alarms() []*etcdserverpb.AlarmMember
+	Range(context.Context, *etcdserverpb.RangeRequest) (*etcdserverpb.RangeResponse, error)
+}
+
+// mux is an interface describing the methods InstallHandler requires.
+type mux interface {
+	Handle(pattern string, handler http.Handler)
+}
+
+type HealthChecker interface {
+	Check() error
+}
+
+type HealthStatus struct {
+	IsOk   bool   `json:"isOk"`
+	Reason string `json:"reason"`
+}
+
+type HealthHandler struct {
```

Comment:
Instead of having one structure to handle all the checks with mirrored code for each type, why not have a generic struct that you can reuse?
```
func InstallHealthEndpoints(mux, server) {
  mux.Handle("/livez", LiveHandler(server))
  mux.Handle("/readyz", ReadyHandler(server))
}

type HealthCheck func() error

type CheckRegistry struct {
  checks map[string]HealthCheck
}

func  LivezHandler() http.Handler {
   reg := CheckRegistry{}
   reg.Register("live", checkAlarms)
   retrurn reg.Handler()
}

func  ReadyzHandler() http.Handler {
   reg := CheckRegistry{}
   reg.Register("alarms", checkAlarms)
   retrurn reg.Handler()
}


func (r *CheckRegister) Register(name string, check HealthCheck) {
  r.checks[name] = check
}

func (r *CheckRegister) Handler() http.Handler {
  return func(_, _) {
  
  }
}
```